File: aws_batch_squared_py/batch_stack.py
"""Batch nested stack — spot/on-demand compute environments, queues, job def.

Ported from ``lib/batch-stack.ts``. Uses the L1 ``Cfn*`` Batch constructs for
full control over allocation strategy and launch-template wiring.
"""

from typing import List

import aws_cdk as cdk
from aws_cdk import aws_batch as batch
from aws_cdk import aws_ec2 as ec2
from aws_cdk import aws_iam as iam
from aws_cdk import aws_logs as logs
from constructs import Construct


class BatchStack(cdk.NestedStack):
    on_demand_queue_arn: str
    spot_queue_arn: str
    security_group: ec2.SecurityGroup
    generic_job_definition_arn: str

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        namespace: str,
        vpc: ec2.IVpc,
        subnets: List[ec2.ISubnet],
        launch_template_id: str,
        batch_service_role: iam.Role,
        instance_profile: iam.InstanceProfile,
        spot_fleet_role: iam.Role,
        on_demand_min_cpus: int,
        on_demand_max_cpus: int,
        spot_min_cpus: int,
        spot_max_cpus: int,
        batch_on_demand_instance_types: str,
        batch_spot_instance_types: str,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create Security Group
        self.security_group = ec2.SecurityGroup(
            self,
            "SecurityGroup",
            vpc=vpc,
            description="SG for genomics workflows on Batch",
            allow_all_outbound=True,
        )

        # Allow all TCP traffic within the security group
        self.security_group.add_ingress_rule(
            self.security_group,
            ec2.Port.all_tcp(),
            "Allow all TCP traffic within security group",
        )

        # Parse instance types
        on_demand_instance_types = (
            ["optimal"]
            if batch_on_demand_instance_types == "optimal"
            else [t.strip() for t in batch_on_demand_instance_types.split(",")]
        )
        spot_instance_types = (
            ["optimal"]
            if batch_spot_instance_types == "optimal"
            else [t.strip() for t in batch_spot_instance_types.split(",")]
        )

        # Create Spot Compute Environment using CfnComputeEnvironment for more control
        spot_compute_env = batch.CfnComputeEnvironment(
            self,
            "SpotComputeEnv",
            compute_environment_name=f"spot-{namespace}-v4",
            service_role=batch_service_role.role_arn,
            type="MANAGED",
            state="ENABLED",
            compute_resources=batch.CfnComputeEnvironment.ComputeResourcesProperty(
                type="SPOT",
                # Optimizes for both price and interruption-resistance (vs plain
                # capacity-optimized).
                allocation_strategy="SPOT_PRICE_CAPACITY_OPTIMIZED",
                ec2_configuration=[
                    batch.CfnComputeEnvironment.Ec2ConfigurationObjectProperty(
                        image_type="ECS_AL2",
                    )
                ],
                launch_template=batch.CfnComputeEnvironment.LaunchTemplateSpecificationProperty(
                    launch_template_id=launch_template_id,
                    version="$Latest",
                ),
                instance_role=instance_profile.instance_profile_arn,
                instance_types=spot_instance_types,
                minv_cpus=spot_min_cpus,
                maxv_cpus=spot_max_cpus,
                security_group_ids=[self.security_group.security_group_id],
                spot_iam_fleet_role=spot_fleet_role.role_arn,
                subnets=[s.subnet_id for s in subnets],
                tags={"Name": f"{namespace}-compute-instance"},
            ),
        )

        # Create On-Demand Compute Environment
        on_demand_compute_env = batch.CfnComputeEnvironment(
            self,
            "OnDemandComputeEnv",
            compute_environment_name=f"ondemand-{namespace}-v4",
            service_role=batch_service_role.role_arn,
            type="MANAGED",
            state="ENABLED",
            compute_resources=batch.CfnComputeEnvironment.ComputeResourcesProperty(
                type="EC2",
                # Progressive fall-through to other instance types when the preferred type
                # has no capacity, instead of wedging jobs in RUNNABLE (the failure mode of
                # plain BEST_FIT).
                allocation_strategy="BEST_FIT_PROGRESSIVE",
                ec2_configuration=[
                    batch.CfnComputeEnvironment.Ec2ConfigurationObjectProperty(
                        image_type="ECS_AL2",
                    )
                ],
                launch_template=batch.CfnComputeEnvironment.LaunchTemplateSpecificationProperty(
                    launch_template_id=launch_template_id,
                    version="$Latest",
                ),
                instance_role=instance_profile.instance_profile_arn,
                instance_types=on_demand_instance_types,
                minv_cpus=on_demand_min_cpus,
                maxv_cpus=on_demand_max_cpus,
                security_group_ids=[self.security_group.security_group_id],
                subnets=[s.subnet_id for s in subnets],
                tags={"Name": f"{namespace}-compute-instance"},
            ),
        )

        # Create Job Queues
        on_demand_queue = batch.CfnJobQueue(
            self,
            "OnDemandQueue",
            job_queue_name=f"OnDemand-{namespace}",
            priority=1000,
            state="ENABLED",
            compute_environment_order=[
                batch.CfnJobQueue.ComputeEnvironmentOrderProperty(
                    order=1,
                    compute_environment=on_demand_compute_env.ref,
                )
            ],
        )

        spot_queue = batch.CfnJobQueue(
            self,
            "SpotQueue",
            job_queue_name=f"Spot-{namespace}",
            priority=1,
            state="ENABLED",
            compute_environment_order=[
                batch.CfnJobQueue.ComputeEnvironmentOrderProperty(
                    order=1,
                    compute_environment=spot_compute_env.ref,
                )
            ],
        )

        # Create CloudWatch Log Group
        logs.LogGroup(
            self,
            "ContainerInstanceLogGroup",
            log_group_name=f"/aws/ecs/container-instance/{namespace}",
            retention=logs.RetentionDays.ONE_WEEK,
        )

        # Create Generic Batch Job Definition
        generic_job_def = batch.CfnJobDefinition(
            self,
            "GenericBatchJobDefinition",
            job_definition_name=f"generic-batch-{namespace}",
            type="container",
            container_properties=batch.CfnJobDefinition.ContainerPropertiesProperty(
                image="amazonlinux",
                vcpus=4,
                memory=16384,
                mount_points=[
                    batch.CfnJobDefinition.MountPointsProperty(
                        source_volume="aws-cli",
                        container_path="/opt/aws-cli",
                        read_only=True,
                    )
                ],
                volumes=[
                    batch.CfnJobDefinition.VolumesProperty(
                        name="aws-cli",
                        host=batch.CfnJobDefinition.VolumesHostProperty(
                            source_path="/opt/aws-cli",
                        ),
                    )
                ],
                environment=[
                    batch.CfnJobDefinition.EnvironmentProperty(
                        name="PATH",
                        value=(
                            "/opt/aws-cli/bin:/usr/local/sbin:/usr/local/bin:"
                            "/usr/sbin:/usr/bin:/sbin:/bin"
                        ),
                    )
                ],
            ),
        )

        self.on_demand_queue_arn = on_demand_queue.attr_job_queue_arn
        self.spot_queue_arn = spot_queue.attr_job_queue_arn
        self.generic_job_definition_arn = generic_job_def.ref

        # Outputs
        cdk.CfnOutput(self, "OnDemandJobQueueArn", value=self.on_demand_queue_arn)
        cdk.CfnOutput(self, "SpotJobQueueArn", value=self.spot_queue_arn)
        cdk.CfnOutput(
            self,
            "BatchSecurityGroupId",
            value=self.security_group.security_group_id,
            description="The Batch Security Group",
        )
        cdk.CfnOutput(
            self,
            "GenericJobDefinitionArn",
            value=self.generic_job_definition_arn,
            description="Generic job definition for running any container",
        )

File: playbook/ways-to-run/patterns/batch-squared/infrastructure/typescript/lib/batch-stack.ts
import * as cdk from "aws-cdk-lib";
import * as batch from "aws-cdk-lib/aws-batch";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import type * as iam from "aws-cdk-lib/aws-iam";
import * as logs from "aws-cdk-lib/aws-logs";
import type { Construct } from "constructs";

export interface BatchStackProps extends cdk.NestedStackProps {
  namespace: string;
  vpc: ec2.IVpc;
  subnets: ec2.ISubnet[];
  launchTemplateId: string;
  batchServiceRole: iam.Role;
  instanceProfile: iam.InstanceProfile;
  spotFleetRole: iam.Role;
  onDemandMinCpus: number;
  onDemandMaxCpus: number;
  spotMinCpus: number;
  spotMaxCpus: number;
  batchOnDemandInstanceTypes: string;
  batchSpotInstanceTypes: string;
}

export class BatchStack extends cdk.NestedStack {
  public readonly onDemandQueueArn: string;
  public readonly spotQueueArn: string;
  public readonly securityGroup: ec2.SecurityGroup;
  public readonly genericJobDefinitionArn: string;

  constructor(scope: Construct, id: string, props: BatchStackProps) {
    super(scope, id, props);

    // Create Security Group
    this.securityGroup = new ec2.SecurityGroup(this, "SecurityGroup", {
      vpc: props.vpc,
      description: "SG for genomics workflows on Batch",
      allowAllOutbound: true,
    });

    // Allow all TCP traffic within the security group
    this.securityGroup.addIngressRule(
      this.securityGroup,
      ec2.Port.allTcp(),
      "Allow all TCP traffic within security group",
    );

    // Parse instance types
    const onDemandInstanceTypes =
      props.batchOnDemandInstanceTypes === "optimal"
        ? ["optimal"]
        : props.batchOnDemandInstanceTypes.split(",").map((t) => t.trim());

    const spotInstanceTypes =
      props.batchSpotInstanceTypes === "optimal"
        ? ["optimal"]
        : props.batchSpotInstanceTypes.split(",").map((t) => t.trim());

    // Create Spot Compute Environment using CfnComputeEnvironment for more control
    const spotComputeEnv = new batch.CfnComputeEnvironment(
      this,
      "SpotComputeEnv",
      {
        computeEnvironmentName: `spot-${props.namespace}-v4`,
        serviceRole: props.batchServiceRole.roleArn,
        type: "MANAGED",
        state: "ENABLED",
        computeResources: {
          type: "SPOT",
          // Optimizes for both price and interruption-resistance (vs plain capacity-optimized).
          allocationStrategy: "SPOT_PRICE_CAPACITY_OPTIMIZED",
          ec2Configuration: [
            {
              imageType: "ECS_AL2",
            },
          ],
          launchTemplate: {
            launchTemplateId: props.launchTemplateId,
            version: "$Latest",
          },
          instanceRole: props.instanceProfile.instanceProfileArn,
          instanceTypes: spotInstanceTypes,
          minvCpus: props.spotMinCpus,
          maxvCpus: props.spotMaxCpus,
          securityGroupIds: [this.securityGroup.securityGroupId],
          spotIamFleetRole: props.spotFleetRole.roleArn,
          subnets: props.subnets.map((s) => s.subnetId),
          tags: {
            Name: `${props.namespace}-compute-instance`,
          },
        },
      },
    );

    // Create On-Demand Compute Environment
    const onDemandComputeEnv = new batch.CfnComputeEnvironment(
      this,
      "OnDemandComputeEnv",
      {
        computeEnvironmentName: `ondemand-${props.namespace}-v4`,
        serviceRole: props.batchServiceRole.roleArn,
        type: "MANAGED",
        state: "ENABLED",
        computeResources: {
          type: "EC2",
          // Progressive fall-through to other instance types when the preferred type has no
          // capacity, instead of wedging jobs in RUNNABLE (the failure mode of plain BEST_FIT).
          allocationStrategy: "BEST_FIT_PROGRESSIVE",
          ec2Configuration: [
            {
              imageType: "ECS_AL2",
            },
          ],
          launchTemplate: {
            launchTemplateId: props.launchTemplateId,
            version: "$Latest",
          },
          instanceRole: props.instanceProfile.instanceProfileArn,
          instanceTypes: onDemandInstanceTypes,
          minvCpus: props.onDemandMinCpus,
          maxvCpus: props.onDemandMaxCpus,
          securityGroupIds: [this.securityGroup.securityGroupId],
          subnets: props.subnets.map((s) => s.subnetId),
          tags: {
            Name: `${props.namespace}-compute-instance`,
          },
        },
      },
    );

    // Create Job Queues
    const onDemandQueue = new batch.CfnJobQueue(this, "OnDemandQueue", {
      jobQueueName: `OnDemand-${props.namespace}`,
      priority: 1000,
      state: "ENABLED",
      computeEnvironmentOrder: [
        {
          order: 1,
          computeEnvironment: onDemandComputeEnv.ref,
        },
      ],
    });

    const spotQueue = new batch.CfnJobQueue(this, "SpotQueue", {
      jobQueueName: `Spot-${props.namespace}`,
      priority: 1,
      state: "ENABLED",
      computeEnvironmentOrder: [
        {
          order: 1,
          computeEnvironment: spotComputeEnv.ref,
        },
      ],
    });

    // Create CloudWatch Log Group
    new logs.LogGroup(this, "ContainerInstanceLogGroup", {
      logGroupName: `/aws/ecs/container-instance/${props.namespace}`,
      retention: logs.RetentionDays.ONE_WEEK,
    });

    // Create Generic Batch Job Definition
    const genericJobDef = new batch.CfnJobDefinition(
      this,
      "GenericBatchJobDefinition",
      {
        jobDefinitionName: `generic-batch-${props.namespace}`,
        type: "container",
        containerProperties: {
          image: "amazonlinux",
          vcpus: 4,
          memory: 16384,
          mountPoints: [
            {
              sourceVolume: "aws-cli",
              containerPath: "/opt/aws-cli",
              readOnly: true,
            },
          ],
          volumes: [
            {
              name: "aws-cli",
              host: {
                sourcePath: "/opt/aws-cli",
              },
            },
          ],
          environment: [
            {
              name: "PATH",
              value:
                "/opt/aws-cli/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            },
          ],
        },
      },
    );

    this.onDemandQueueArn = onDemandQueue.attrJobQueueArn;
    this.spotQueueArn = spotQueue.attrJobQueueArn;
    this.genericJobDefinitionArn = genericJobDef.ref;

    // Outputs
    new cdk.CfnOutput(this, "OnDemandJobQueueArn", {
      value: this.onDemandQueueArn,
    });

    new cdk.CfnOutput(this, "SpotJobQueueArn", {
      value: this.spotQueueArn,
    });

    new cdk.CfnOutput(this, "BatchSecurityGroupId", {
      value: this.securityGroup.securityGroupId,
      description: "The Batch Security Group",
    });

    new cdk.CfnOutput(this, "GenericJobDefinitionArn", {
      value: this.genericJobDefinitionArn,
      description: "Generic job definition for running any container",
    });
  }
}
